Update continuous aggregate API functions (#496)
timescale/docs.timescale.com-content: api.md

|---|---|---|
|`timescaledb.continuous`|||
|**Description**|**Type**|**Default**|
|If timescaledb.continuous is not specified, then this is a regular PostgresSQL materialized view. | `BOOLEAN` ||

#### Optional `WITH` clause options [](continuous_aggregate-create_view-with-optional)

|**Name**|||
|---|---|---|
|`timescaledb.materialized_only`|||
|**Description**|**Type**|**Default**|

> | Return only materialized data when querying the continuous aggregate view. | `BOOLEAN` | false |

Reviewer: Should this section have any pointer to [real-time aggregates][real-time-aggregates] ?  I don't think the reader just looking up this API spec will understand this term.  If so, where should the pointer go?

Author: Added a pointer to the section on real-time aggregates.

> | Return only materialized data when querying the continuous aggregate view. See more in section on [real-time aggregates][real-time-aggregates]. | `BOOLEAN` | false |

Reviewer: Have you done a local build, and does the visual format support such a wide line in table?

Author: Nope, I have not done a local build, but there's been quite long lines elsewhere.

|   |   |   |
|`timescaledb.create_group_indexes`|||
|**Description**|**Type**|**Default**|
| Create indexes on the materialization table for the group by columns (specified by the `GROUP BY` clause of the `SELECT` query). | `BOOLEAN` | Indexes are created by default for every group by expression + time_bucket expression pair.|

> #### Notes

Reviewer: @solugebefola Do you have thoughts on any standard template format for API docs with such type of "notes"?

- The view will be automatically refreshed (as outlined under
  [`refresh_continuous_aggregate`](#refresh_continuous_aggregate))
  unless `WITH NO DATA` is given (`WITH DATA` is the default).
- `SELECT` query should be of the form specified in the syntax above.
- The hypertable used in the `SELECT` may not have [row-level-security
  policies][postgres-rls] enabled.
-  `GROUP BY` clause must include a time_bucket expression. The
   [`time_bucket`](#time_bucket) expression must use the time
   dimension column of the hypertable.

> - [`time_bucket_gapfill`](#time_bucket_gapfill) is not allowed in
>   continuous aggs, but may be run in a `SELECT` from the continuous
>   aggregate view.

Reviewer: ```suggestion
- [`time_bucket_gapfill`](#time_bucket_gapfill) is not allowed in
  continuous aggs, but may be executed by a `SELECT` on the continuous
  aggregate view.
```

- In general, aggregates which can be [parallelized by
  PostgreSQL][postgres-parallel-agg] are allowed in the view
  definition, this includes most aggregates distributed with
  PostgreSQL. Aggregates with `ORDER BY`, `DISTINCT` and `FILTER`
  clauses are not permitted.
- All functions and their arguments included in `SELECT`, `GROUP BY`
  and `HAVING` clauses must be [immutable][postgres-immutable].

> - Queries with `ORDER BY` are disallowed.

Reviewer: Minor nit:  What's the difference between these line:

     Queries with `ORDER BY` are disallowed.

And the above:

    Aggregates with `ORDER BY`, `DISTINCT` and `FILTER` clauses are not permitted.

Author: Seems like it was accidentally duplicated. Removed the second one.

#### Sample Usage [](continuous_aggregate-create-examples)
Create a continuous aggregate view.
```sql
CREATE MATERIALIZED VIEW continuous_aggregate_view( timec, minl, sumt, sumh )
WITH (timescaledb.continuous) AS
    SELECT time_bucket('1day', timec), min(location), sum(temperature), sum(humidity)
        FROM conditions

>         GROUP BY time_bucket('1day', timec)

Reviewer: I would standardize on 2 space indents.  The section actually has a mix (this is old issue =)

Author: I reformatted this code block and the next code block.

REFRESH MATERIALIZED VIEW contagg_view;
```

---

## DROP MATERIALIZED VIEW (Continuous Aggregate) :community_function: [](continuous_aggregate-drop_view)
Continuous aggregate views can be dropped using the `DROP MATERIALIZED VIEW` statement.

> This statement deletes the continuous aggregate and all its internal objects. To also delete other dependent objects, add the `CASCADE` option. Dropping a continuous aggregate does not affect the data in the

Reviewer: Can you give an example of what dependent objects might be?

"To also delete other dependent objects (e.g., <...>), add the ..."

Author: Added an example as subphrase.

REFRESH MATERIALIZED VIEW contagg_view;
```

---

## DROP MATERIALIZED VIEW (Continuous Aggregate) :community_function: [](continuous_aggregate-drop_view)
Continuous aggregate views can be dropped using the `DROP MATERIALIZED VIEW` statement.

This statement deletes the continuous aggregate and all its internal

> objects. To also delete other dependent objects&mdash;such as a view

Reviewer: Does `&mdash;` work in the markdown parser we are using?

Author: I don't know for sure, but this is what the reference says (https://daringfireball.net/projects/markdown/syntax#html):
> For any markup that is not covered by Markdown’s syntax, you simply use HTML itself. There’s no need to preface it or delimit it to indicate that you’re switching from Markdown to HTML; you just use the tags.

#### Sample Usage [](continuous_aggregate-drop_view-examples)
Drop existing continuous aggregate.
```sql
DROP MATERIALIZED VIEW contagg_view;
```

## refresh_continuous_aggregate() :community_function: [](refresh_continuous_aggregate)

> Refresh all buckets of a continuous aggregate between two points of
> time. Note that the function expects the parameter values to have the same time type as used in the continuous aggregate's time bucket expression.

Reviewer: ```suggestion
Refresh all buckets of a continuous aggregate between two points of
time. 

The function expects the parameter values to have the same time type
as used in the continuous aggregate's time bucket expression (e.g., if
the time bucket specifies in `timestamptz`, then the start and end time
supplied should also be `timestamptz`).
```

#### Sample Usage [](continuous_aggregate-drop_view-examples)
Drop existing continuous aggregate.
```sql
DROP MATERIALIZED VIEW contagg_view;
```

## refresh_continuous_aggregate() :community_function: [](refresh_continuous_aggregate)

Refresh all buckets of a continuous aggregate inside a window. Note

> that only buckets fully in the window will be refreshed, so if any

Author: Removed it.

buckets are partially outside the window either at the beginning or
the end will be excluded from the refresh.

#### Required Arguments [](refresh_continuous_aggregate-required-arguments)

|Name|Description|
|---|---|
| `continuous_aggregate` | (REGCLASS) The continuous aggregate to refresh. |
| `window_start` | (INTERVAL) Start of the window relative to current time to refresh. |

> | `window_end` | (INTERVAL) End of the window relative to current time to refresh. |

Reviewer 2: I also think we need to describe somewhere what reasonable values for these parameters are. For instance, window_start should be before window_end. And, schedule_interval should (must) be smaller than window start. Perhaps for a later PR, but I am worried that this will be lost.

> | `window_end` | (INTERVAL) End of the window relative to current time to refresh. |

Author: I have deliberately skipped discussing reasonable values for the parameters here since it is not clear what we should recommend. I think we can handle it in a later PR. I added a note about that to the documentation issue.

> | `window_end` | (INTERVAL) End of the window relative to current time to refresh. |

Reviewer 2: I am not talking about recommended "defaults", but rather configuration that doesn't make sense. For instance, for the policy to make sense:

* window_start should be larger than window_end
* schedule_interval should be smaller than window_start

and so on.

> | `window_end` | (INTERVAL) End of the window relative to current time to refresh. |

Author: I added a notice about this to the `add_continuous_aggregate_policy` but I do not understand how this affects `refresh_continuous_aggregate`. It might not even be the case that there is a policy defined.

#### Sample Usage [](continuous_aggregate-drop_view-examples)
Drop existing continuous aggregate.
```sql
DROP MATERIALIZED VIEW contagg_view;
```

## refresh_continuous_aggregate() :community_function: [](refresh_continuous_aggregate)

Refresh all buckets of a continuous aggregate between two points of
time. 

The function expects the parameter values to have the same time type
as used in the continuous aggregate's time bucket expression (e.g., if
the time bucket specifies in `timestamptz`, then the start and end time
supplied should also be `timestamptz`).

#### Required Arguments [](refresh_continuous_aggregate-required-arguments)

|Name|Description|
|---|---|
| `continuous_aggregate` | (REGCLASS) The continuous aggregate to refresh. |
| `window_start` | Start of the window to refresh, has to be before `window_end`. |
| `window_end` | End of the window to refresh, has to be after `window_start`. |

> 

Reviewer: Any sample usage?

Author: Added one.

## add_continuous_aggregate_policy() :community_function> [](add_continuous_aggregate_policy)

Create a policy that automatically refreshes a continuous aggregate.

#### Required Arguments [](add_continuous_aggregate_policy-required-arguments)

|Name|Description|
|---|---|
| `continuous_aggregate` | (REGCLASS) The continuous aggregate to add the policy for. |
| `start_interval` | (INTERVAL) Start of the refresh window relative to the time when the policy is executed. |
| `end_interval` | (INTERVAL) End of the refresh window relative to the time when the policy is executed. |
| `schedule_interval` | (INTERVAL) Interval between refresh executions. |

#### Optional Arguments [](add_continuous_aggregate_policy-optional-arguments)

|Name|Description|
|---|---|
| `if_not_exists` | (BOOLEAN) Set to true to avoid throwing an error if the continuous aggregate policy already exists. A notice is issued instead. Defaults to false. |

#### Returns [](add_continuous_aggregate_policy-returns)

|Column|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id created to implement this policy|

#### Sample Usage [](add_continuous_aggregate_policy-examples)

> To add a policy that refreshes the last month once an hour:

Reviewer 2: ```suggestion
Add a policy that refreshes the last month once an hour, excluding the latest hour from the aggregate (for performance reasons, it is recommended to exclude buckets that still see lots of writes):
```

The above can be complemented by linking to a deeper discussion on recommended settings.

performance on older chunks. Each policy is assigned a scheduled job
which will be run in the background to enforce it.

## add_continuous_aggregate_policy() :community_function> [](add_continuous_aggregate_policy)

> Create a policy that refresh a windows in a continuous aggregate.

Reviewer 2: ```suggestion
Create a policy that automatically refreshes a continuous aggregate.
```

#### Required Arguments [](add_continuous_aggregate_policy-required-arguments)

|Name|Description|
|---|---|
| `continuous_aggregate` | (REGCLASS) The continuous aggregate to add the policy for. |

> | `start_interval` | (INTERVAL) Start of the window relative to current time to refresh. |
> | `end_interval` | (INTERVAL) End of the window relative to current time to refresh. |

Reviewer 2: ```suggestion
| `start_interval` | (INTERVAL) Start of the refresh window relative to the time when the policy is executed. |
| `end_interval` | (INTERVAL) End of the refresh window relative to the time when the policy is executed. |
```

| `schedule_interval` | (INTERVAL) Interval between refresh executions. |

#### Optional Arguments [](add_continuous_aggregate_policy-optional-arguments)

|Name|Description|
|---|---|

> | `if_not_exists` | (BOOLEAN) Set to true to avoid throwing an error if the continuous aggregate_policy already exists. A notice is issued instead. Defaults to false. |

Reviewer 2: ```suggestion
| `if_not_exists` | (BOOLEAN) Set to true to avoid throwing an error if the continuous aggregate policy already exists. A notice is issued instead. Defaults to false. |
```

#### Returns [](add_continuous_aggregate_policy-returns)

|Column|Description|
|---|---|
|`job_id`| (INTEGER)  TimescaleDB background job id created to implement this policy|

#### Sample Usage [](add_continuous_aggregate_policy-examples)

To add a policy that refreshes the last month once an hour:
```sql
SELECT add_continuous_aggregate_policy('conditions_summary',
	start_interval => INTERVAL '1 month',

> 	end_interval => INTERVAL '1 hour',

Reviewer 2: The setting of end_interval is not motivated. And we never describe why you'd want to have a refresh offset/lag like this.

Author: Not sure what you mean. All the parameters given here are mandatory and have to be given.

Reviewer 2: Above you say `add a policy that refreshes the last month once an hour`, which isn't strictly true since you have an "offset" of 1 hour via the `end_interval` (which takes 1 hour off of the month). This begs the question: "why is there an offset of 1 hour?". We never explain in the API documentation or the example why you'd like to have this 1 hour offset, so it just comes out of nowhere. It is there to avoid write amplification, but no such context or explanation is given.

The other parameters are explained, self-evident, or somewhat motivated by the text and given context.

Author: It is not clear if you want:
* To rewrite the documentation and add a discussion about the write amplification. I have deliberately skipped this because 1) it requires a discussion about recommended values and what yo use, which is not obvious, and 2) it is not suitable material for the API reference documentation and is more suitable for the "Using" section.
* To remove `end_interval`, which is not possible because it is mandatory to the function.
* Set `end_interval` to NULL, which raises a discussion about the write amplification again, and the API documentation is not a suitable place to discuss such things.

Reviewer: (Not sure exactly what @erimatnor is proposing.  Agree that longer discussion should be in a "Using" section.  Why don't we revisit after we have a draft of that section, to see how we might reference it?)

Reviewer 2: @mfreed my suggested change was already adopted as part of another comment, which is no longer visible.

performance on older chunks. Each policy is assigned a scheduled job
which will be run in the background to enforce it.

## add_continuous_aggregate_policy() :community_function> [](add_continuous_aggregate_policy)

Create a policy that automatically refreshes a continuous aggregate.

#### Required Arguments [](add_continuous_aggregate_policy-required-arguments)

|Name|Description|
|---|---|
| `continuous_aggregate` | (REGCLASS) The continuous aggregate to add the policy for. |
| `start_interval` | Start of the refresh window as an interval relative to the time when the policy is executed. |
| `end_interval` | End of the refresh window as an interval relative to the time when the policy is executed. |

> | `schedule_interval` | Interval between refresh executions in wall-clock time. |

Reviewer: Question:  was there discussion to make `schedule_interval` an optional argument, default to to the time_bucket of the continuous agg?

Author: There has been discussions about if any of these parameters should be made optional, but for the time being, they are mandatory. If it changes, we have to change the documentation.
